File: 2487-remove-nodes-from-linked-list/2487-remove-nodes-from-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNodes(ListNode head) {
        if(head == null || head.next == null) return head;
        
        ListNode reversedHead = reverse(head);
        ListNode ptr = reversedHead;
        ListNode prev = null;
        int max = reversedHead.val;
        while(ptr!=null){
            if(max <= ptr.val){
                // update max
                max = ptr.val;
                prev = ptr;
            }else{
                // remove ptr node
                prev.next = ptr.next;
            }
            ptr = ptr.next;
        }
        return reverse(reversedHead);
    }
    
    private ListNode reverse(ListNode head){
        ListNode ptr = null;
        ListNode left = null;
        ListNode right = head;
        
        while(right!=null){
            ptr = right.next;
            right.next = left;
            left = right;
            right = ptr;
        }
        return left;
    }
    
}

File: 2487-remove-nodes-from-linked-list/2487-remove-nodes-from-linked-list.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun removeNodes(head: ListNode?): ListNode? {
        head ?: return null
        val stack = Stack<Int>()

        var ptr = head
        while(ptr!=null){
            stack.add(ptr.`val`)
            ptr = ptr.next
        }
        
        var max = stack.peek()
        var prev : ListNode? = null
        var curr : ListNode? = null
        while(!stack.isEmpty()){
            val temp = stack.pop()
            if(max <= temp){
                max = temp
                curr = ListNode(temp)
                curr.next = prev
                prev = curr
            }
        }
        
        return prev
    }
}